selene-lib/src/rules/limit_function_complexity.rs

```diff
@@ -0,0 +1,272 @@
+use super::*;
+use crate::ast_util::range;
+use std::convert::Infallible;
+
+
+use full_moon::{
+    ast::{self, Ast},
+    // node::Node,
+    visitors::Visitor,
+};
+
+use serde::Deserialize;
+
+#[derive(Clone, Copy, Deserialize)]
+pub struct LimitFunctionComplexityConfig {
+    maximum_complexity: u16,
+}
+
+impl Default for LimitFunctionComplexityConfig {
+    fn default() -> Self {
+        Self {
+            // eslint defaults to 20, but testing on OSS Lua shows that 20 is too aggressive
+            maximum_complexity: 40.to_owned(),
+        }
+    }
+}
+
+
+#[derive(Default)]
+pub struct LimitFunctionComplexityLint {
+    config: LimitFunctionComplexityConfig
+}
+
+impl Rule for LimitFunctionComplexityLint {
+    type Config = LimitFunctionComplexityConfig;
+    type Error = Infallible;
+
+    const SEVERITY: Severity = Severity::Warning;
+    const RULE_TYPE: RuleType = RuleType::Style;
+
+    fn new(config: Self::Config) -> Result<Self, Self::Error> {
+        Ok(LimitFunctionComplexityLint { config })
+    }
+
+    fn pass(&self, ast: &Ast, _: &Context, _: &AstContext) -> Vec<Diagnostic> {
+        let mut visitor = LimitFunctionComplexityVisitor {
+            positions: Vec::new(),
+            config: self.config,
+        };
+
+        visitor.visit_ast(ast);
+
+        visitor
+            .positions
+            .into_iter()
+            .map(|position| {
+                Diagnostic::new(
+                    "limit_function_complexity",
+                    format!(
+                        "function has cyclomatic complexity of {}, exceeding configured limit of {}",
+                        position.1,
+                        self.config.maximum_complexity
+                    ).to_owned(),
+                    Label::new(position.0),
+                )
+            })
+            .collect()
+    }
+}
+
+struct LimitFunctionComplexityVisitor {
+    positions: Vec<((u32, u32), u16)>,
+    config: LimitFunctionComplexityConfig,
+}
+
+fn count_expression_complexity(expression: &ast::Expression, starting_complexity: u16) -> u16 {
+    let mut complexity = starting_complexity;
+
+    match expression {

```

Comment:
```suggestion
    #[cfg_attr(
        feature = "force_exhaustive_checks",
        deny(non_exhaustive_omitted_patterns)
    )]
    match expression {
```

This will give a compile error in tests when a new expression is added, but not accounted for yet.